# Исследование надежности заемщиков

## Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Описание данных
- `children` — количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита

## Оглавление

1. Предобработка данных
2. Исследование данных

# Предобработка данных

In [1]:
# импортируйте библиотеку pandas
import pandas as pd

In [2]:
try:
    data = pd.read_csv('data.csv')
except FileNotFoundError:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv') 

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Удаление пропусков

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform("median"))
    
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

## Обработка аномальных значений

In [7]:
data['days_employed'] = data.loc[data['days_employed'] < 0, 'days_employed'] = abs(data['days_employed'])

In [8]:
print(data.groupby('income_type')['days_employed'].median())

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64


In [9]:
print(data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


In [10]:
data = data.drop(data[(data['children'] == -1) | (data['children'] == 20)].index)

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

## Удаление пропусков (продолжение)

In [12]:
data['days_employed'] = data['days_employed'].fillna(data.groupby('income_type')['days_employed'].transform("median"))

In [13]:
data.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

## Изменение типов данных

In [14]:
data['total_income'] = data['total_income'].astype(int)

## Обработка дубликатов

In [15]:
data['education'] = data['education'].str.lower()

In [16]:
print(data.duplicated().sum())

71


In [17]:
data = data.drop_duplicates().reset_index(drop=True)

## Категоризация данных

In [18]:
def categorize_income(income):
    if income in range(0, 30001):
        return 'E'
    elif income in range(30001, 50001):
        return 'D'
    elif income in range(50001, 200001):
        return 'C'
    elif income in range(200001, 1000001):
        return 'B'
    elif income > 1000001:
        return 'A'

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [25]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [26]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

# Исследование данных данные

In [27]:
def proportion(pdSerises):
    """возвращаем отношение суммы невыплативших долг ко всем кто его взял, 
    выраженное в процентах, округленное до двух знаков"""
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 2)) + '%' 

In [28]:
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

Большинство заемщиков имеет не более трех детей, поэтому сделаем 5 групп:

* Нет детей
* Один ребенок
* Два ребенка
* Три ребенка
* Много детей

Для этого реализуем функцию children_category()

In [29]:
def children_category(quantity):
    if quantity == 0:
        return 'Нет детей'
    elif quantity == 1:
        return 'Один ребенок'
    elif quantity == 2:
        return 'Два ребенка'
    elif quantity == 3:
        return 'Три ребенка'
    else:
        return 'Много детей'

In [30]:
data['children_category'] = data['children'].apply(children_category)

Для удобства и наглядность изысканий создадим сводную таблицу методом .pivot_table()

In [31]:
data_pivot = data.pivot_table(index=['children_category'], values=['debt'], aggfunc=['sum', 'count', proportion]) # 'sum' - количество людей из соответсвующей категории которые не вернули долг вовремя. 'count' - общее количество должников в данной категории. 'propotion' - отношении 'sum' к 'count'
data_pivot = data_pivot.sort_values(by = ('proportion', 'debt'))
display(data_pivot)

sum  count proportion
                   debt   debt       debt
children_category                        
Нет детей          1063  14091      7.54%
Много детей           4     50       8.0%
Три ребенка          27    330      8.18%
Один ребенок        444   4808      9.23%
Два ребенка         194   2052      9.45%

## Промежуточный вывод

Вывод в данном вопросе весьма неоднозначный, проведенные расчеты показывают, что заемщики у которых нет детей чуть чаще остальных возвращают кредит вовремя. Однако стоит заметить, что данные по другим категориям заемщиков выше не более чем на 2%, что говорит нам о том, что прямой зависимости между количеством детей и фактом несвоевременного погащения долга нет.

In [32]:
data_pivot = data.pivot_table(index=['family_status'], values=['debt'], aggfunc=['sum', 'count', proportion]) # 'sum' - количество людей из соответсвующей категории которые не вернули долг вовремя. 'count' - общее количество должников в данной категории. 'propotion' - отношении 'sum' к 'count'
data_pivot = data_pivot.sort_values(by = ('proportion', 'debt'))
display(data_pivot)

sum  count proportion
                      debt   debt       debt
family_status                               
вдовец / вдова          63    951      6.62%
в разводе               84   1189      7.06%
женат / замужем        927  12261      7.56%
гражданский брак       385   4134      9.31%
Не женат / не замужем  273   2796      9.76%

## Промежуточный вывод

Проведенный анализ заемщиков по данному кретерию выявил, что наиболее ответственными являются люди, потерявщие свою вторую половинку(вдовец/вдова). Заемщики из категории 'в разводе' и 'женат / замужем' находятся на одинаковом уровне опасности для банка. Наиболее опасными для банка являются одинокие люди, не состоящие в отношениях из группы 'Не женат / не замужем' они с большей вероятностью не вернут кредит.

## Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ранее мы уже провели категоризацию заемщиков по уровню дохода, присвоив им следующие категории:

* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Необходимо проанализировать клиентов банка по этому кретерию.

In [33]:
data_pivot = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum', 'count', proportion]) # 'sum' - количество людей из соответсвующей категории которые не вернули долг вовремя. 'count' - общее количество должников в данной категории. 'propotion' - отношении 'sum' к 'count'
data_pivot = data_pivot.sort_values(by = ('proportion', 'debt'))
display(data_pivot)

sum  count proportion
                       debt   debt       debt
total_income_category                        
D                        21    349      6.02%
B                       354   5014      7.06%
A                         2     25       8.0%
C                      1353  15921       8.5%
E                         2     22      9.09%

### Dывод

Результаты исследования по данному вопросы немного удивили. Наиболее ответственными оказались заемщики из класса 'D' со сравнительно небольшим доходом. Наиболее рискованной группой ожидаемо стали люди с минимальным заработком из класса 'E'. Однако  группа 'E' представленна малым количеством заемщиков отчего неменее рискованной является самая многочисленная группа 'C' с результатом невозврата долга в 8.5%.

## Как разные цели кредита влияют на его возврат в срок?

In [35]:
data_pivot = data.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum', 'count', proportion]) # 'sum' - количество людей из соответсвующей категории которые не вернули долг вовремя. 'count' - общее количество должников в данной категории. 'propotion' - отношении 'sum' к 'count'
data_pivot = data_pivot.sort_values(by = ('proportion', 'debt'))
display(data_pivot)

sum  count proportion
                         debt   debt       debt
purpose_category                               
операции с недвижимостью  780  10751      7.26%
проведение свадьбы        183   2313      7.91%
получение образования     369   3988      9.25%
операции с автомобилем    400   4279      9.35%

### Вывод

Операции с недвижимостью - 7.26% Заемщики, бравшие в долг по данной причине с наибольшей вероятностью среди остальных групп вернут долг банку.

Проведение свадьбы - 7.91% Сравнительно безопасная группа заемщиков, бравших долг на проведение торжества влюбленных. Вероятно низкий процент связан со спецификой торжества, помощью родственников и небольшой суммой долга.

Получение образования - 9.25% Одна из наиболее опасных категорий заемщиков. Вероятно в современном развивающимся мире сложно найти актуальное предложение в сфере образования, где за короткий срок можно освоить востребованную специальность, если это конечно, не Yandex Practicum :)

Операции с автомобилем  - 9.35% Автомобиль - средство повышенной опасности, заемщики не застрахованы от непредвиденных ситуаций, аварий по вине заемщика итд. Также автомобиль быстро теряет в стоимости по сравнению с постоянно растущей стоимостью недвижимости.

## Возможные причины появления пропусков в исходных данных?

Вероятно возникновение большинства пропусков можно обьяснить тем, что большое количество заемщиков может работать в 'Черную' или в 'Серую' не имея официального дохода, работая без трудовых договоров не накапливая стаж.
Также нельзя исключать 'человеческий фактор' который неизменно приводит к ошибкам, тем больше, чем больше обьем данных.

# Общий вывол
По итогу исследований мы выявили наилучшего заемщика, это - человек без детей, состоящий или состоявший в официальных отношениях, имеющий уровень дохода выше среднего и берущий деньги на недвижимость.

Анализ по каждому кретерию выявил, что разброс несвоеевременной выплаты долга в среднем коллеблится от 7% до 9% соответсвенно.

* Влияние детей
Нет детей - 7.54% \
Один ребенок - 9.23%\
Два ребенка - 9.45%\
Три ребенка - 8.18%\
Много детей - 8.0%


* Влияние семейного положения
вдовец / вдова - 6.62%\
в разводе - 7.06%\
женат / замужем - 7.56%\
гражданский брак - 9.31%\
Не женат / не замужем - 9.76%


* Влияние уровня дохода
0–30000 — 9.09%\
30001–50000 — 6.02%\
50001–200000 — 8.5\
200001–1000000 — 7.06%\
1000001 и выше — 8.0%


* Влияние цели кредита
операции с недвижимостью - 7.26%\
проведение свадьбы - 7.91%\
получение образования - 9.25%\
операции с автомобилем - 9.35%


**Замечание:**

Результаты, полученные по признаку 'уровень дохода' стоит подвергнуть дополнительному анализу из-за недостаточности данных в категориях людей с минимальным доходом и людей со сверх доходом. Суммарно людей в этих двух категориях всего 49 человек, что на составляет 0.2% от общего числа заемщиков.